tests/unit/test_business_logic.py

```diff
@@ -1441,6 +1442,201 @@ def test_add_file_to_request_with_filetype(bll, filetype, success):
+    # the correct context to hit this test
+
+    workspace = bll.get_workspace(release_request.workspace, author)
+    src = workspace.abspath(relpath)
+    file_id = store_file(release_request, src)
+    manifest = workspace.get_manifest_for_file(relpath)
+    group_name = "default"
+
+    audit = AuditEvent.from_request(
+        request=release_request,
+        type=AuditEventType.REQUEST_FILE_ADD,
+        user=author,
+        path=relpath,
+        group=group_name,
+        filetype=RequestFileType.OUTPUT.name,
+    )
+
+    bll._dal.add_file_to_request(
+        request_id=release_request.id,
+        group_name=group_name,
+        relpath=relpath,
+        file_id=file_id,
+        filetype=RequestFileType.OUTPUT,
+        timestamp=manifest["timestamp"],
+        commit=manifest["commit"],
+        repo=manifest["repo"],
+        size=manifest["size"],
+        job_id=manifest["job_id"],
+        row_count=manifest["row_count"],
+        col_count=manifest["col_count"],
+        audit=audit,
+    )
+    with pytest.raises(
+        bll.RequestPermissionDenied, match=r"Cannot update file of type"
+    ):
+        bll.update_file_in_request(release_request, relpath, author)
+
+
+def test_update_file_in_request_not_updated(bll):
+    author = factories.create_user("author", ["workspace"], False)
+
+    relpath = UrlPath("path/file.txt")
+    workspace = factories.create_workspace("workspace")
+    factories.write_workspace_file(workspace, relpath)
+    release_request = factories.create_request_at_status(
+        "workspace",
+        author=author,
+        status=RequestStatus.RETURNED,
+    )
+
+    with pytest.raises(bll.RequestPermissionDenied, match=r"not updated"):
+        bll.update_file_in_request(release_request, relpath, author)
+
+
+@pytest.mark.parametrize(
+    "status,reviewed,approved,success,notification_sent",
+    [
+        (RequestStatus.PENDING, False, None, True, False),
+        (RequestStatus.SUBMITTED, False, None, False, False),
+        (RequestStatus.WITHDRAWN, True, False, False, False),
+        (RequestStatus.APPROVED, True, True, False, False),
+        (RequestStatus.REJECTED, True, False, False, False),
+        (RequestStatus.PARTIALLY_REVIEWED, True, True, False, False),
+        (RequestStatus.REVIEWED, True, True, False, False),
+        (RequestStatus.RETURNED, True, False, True, True),
+        (RequestStatus.RELEASED, True, True, False, False),
+    ],
+)
+def test_update_file_to_request_states(
+    status, reviewed, approved, success, notification_sent, bll, mock_notifications
+):
+    author = factories.create_user("author", ["workspace"], False)
+    checkers = factories.get_default_output_checkers()
+    workspace = factories.create_workspace("workspace")
+    path = UrlPath("path/file.txt")
+
+    if status == RequestStatus.WITHDRAWN:
+        withdrawn_after = RequestStatus.RETURNED
+    else:
+        withdrawn_after = None
+
+    if reviewed:
```

Comment:
In `request_file`, approved=True means that it's approved by 2 reviewers, rejected=True means that it's rejected by 2 reviewers (and it's an if/elif, so if you call it with approved=True and rejected=True, it'll be approved.). So I think you can get rid of the reviewed parameter here, and get the same file with:
```
workspace_file = factories.request_file(
    path=path,
    user=author,
    approved=approved,
    rejected=rejected
)
```

Reply:
good idea, done